File: cores/Rust/compositron/src/importers/meps_dat_importer.rs
use std::fs::File;
use std::io::{BufRead, Read};
use std::path::{Path, PathBuf};

use thiserror::Error;

use crate::core::Measurement;
use crate::pals::PALSpectrum;
use crate::dbs::DBSpectrum;
use crate::core::utils::Spectrum;

#[derive(Debug, Error)]
pub enum DATFormatError {
    #[error("Invalid spectrum format")]
    SpectrumParseError {
        #[from]
        inner: std::num::ParseIntError,
    },

    #[error("Invalid header field: {field}")]
    HeaderParseError {
        #[source]
        inner: std::num::ParseFloatError,
        field: String,
    },

    #[error("{info}")]
    HeaderFormatError {
        info: String,
    }
}

#[derive(Debug, Error)]
pub enum ImportError {
    #[error("IO Error opening {path}")]
    FileSystemError {
        #[source]
        inner: std::io::Error,
        path: PathBuf,
    },

    #[error("Invalid MePS file format")]
    InvalidMePSFormatError {
        #[from]
        inner: DATFormatError,
    }
}

struct MePSFileBundle {
    spectrum_file: String,
    param_file: String,
    start_phs_file: String,
    stop_phs_file: String,
}

struct HeaderInfo {
    index: u32,
    sample: String,
    energy: f64,
    temperature: f64,
    bin_width: f64,
}

fn split_filepath_componentes<'a>(filepath: &'a str) -> Option<Vec<&'a str>> {
    let provided_path = Path::new(filepath);
    let provided_filename = provided_path.file_name()?.to_str()?;
    let filename_components: Vec<&str> = provided_filename.split("_").collect();
    Some(filename_components)
}

fn get_all_filenames(filepath: &str) -> Result<MePSFileBundle, ImportError> {
    let filename_components = split_filepath_componentes(filepath).ok_or(
        ImportError::FileSystemError {
            inner: std::io::Error::from(std::io::ErrorKind::InvalidFilename),
            path: filepath.into(),
        }
    )?;

    let critical_idx = filename_components.len() - 3;

    match filename_components[critical_idx] {
        "PALS" => {
            let begin = filename_components[..critical_idx].join("_");
            let end = filename_components[critical_idx+1..].join("_");
            Ok(MePSFileBundle {
                spectrum_file: filename_components.join("_"),
                param_file: [&begin, "Measurement_Parameter", &end].join("_"),
                start_phs_file: [&begin, "PHS_Ch0", &end].join("_"),
                stop_phs_file: [&begin, "PHS_Ch1", &end].join("_")
            })
        },
        "Parameter" => {
            let begin = filename_components[..critical_idx-1].join("_");
            let end = filename_components[critical_idx+1..].join("_");
            Ok(MePSFileBundle {
                spectrum_file: [&begin, "PALS", &end].join("_"),
                param_file: filename_components.join("_"),
                start_phs_file: [&begin, "PHS_Ch0", &end].join("_"),
                stop_phs_file: [&begin, "PHS_Ch1", &end].join("_")
            })
        },
        "Ch0" => {
            let begin = filename_components[..critical_idx-1].join("_");
            let end = filename_components[critical_idx+1..].join("_");
            Ok(MePSFileBundle {
                spectrum_file: [&begin, "PALS", &end].join("_"),
                param_file: [&begin, "Measurement_Parameter", &end].join("_"),
                start_phs_file: filename_components.join("_"),
                stop_phs_file: [&begin, "PHS_Ch1", &end].join("_")
            })
        },
        "Ch1" => {
            let begin = filename_components[..critical_idx-1].join("_");
            let end = filename_components[critical_idx+1..].join("_");
            Ok(MePSFileBundle {
                spectrum_file: [&begin, "PALS", &end].join("_"),
                param_file: [&begin, "Measurement_Parameter", &end].join("_"),
                start_phs_file: [&begin, "PHS_Ch0", &end].join("_"),
                stop_phs_file: filename_components.join("_")
            })
        },
        _ => {
            Err(ImportError::FileSystemError {
                inner: std::io::Error::from(std::io::ErrorKind::InvalidFilename),
                path: filepath.into(),
            })
        },
    }
}

fn parse_header_field(
    field: &str, expected_unit: &str
) -> Result<f64, DATFormatError> {
    let l = field.len();
    let u = expected_unit.len();

    let value = field[..l-u].parse::<f64>().map_err(|err|
        DATFormatError::HeaderParseError {
            inner: err, field: field[..l-u].into()
        },
    )?;
    let unit = &field[l-u..];

    if unit == expected_unit {
        Ok(value)
    } else {
        Err(DATFormatError::HeaderFormatError {
            info: format!(
                "Unexpected unit {unit} in spectrum file header \
                (expected {expected_unit})"
            )
        })
    }
}

fn get_header_info(
    header: &str, m: &mut Measurement
) -> Result<HeaderInfo, DATFormatError> {
    let header_fields: Vec<&str> = header
        .split("_")
        .map(|field| field.trim())
        .collect();

    let header_size = header_fields.len();

    if header_size < 6 {
        return Err(DATFormatError::HeaderFormatError {
            info: "Spectrum header has unexpected format".into()
        });
    }

    let index = header_fields[0].parse::<u32>()?;
    let bin_width = parse_header_field(header_fields[header_size-1], "ns")?;
    let temperature = parse_header_field(header_fields[header_size-3], "K")?;
    let energy = parse_header_field(header_fields[header_size-4], "keV")?;
    let sample = header_fields[1..header_size-5].join("_");

    m.metadata.insert("Sample".into(), sample.clone());
    m.metadata.insert("PositronImplantationEnergy".into(), format!("{energy}"));
    m.metadata.insert("SampleTemperature".into(), format!("{temperature}"));

    Ok(HeaderInfo {
        index,
        sample,
        energy,
        temperature,
        bin_width,
    })
}

fn compare_filename_and_header(filepath: &PathBuf, header: &HeaderInfo) {
    let Some(filename) = filepath.file_name() else { return; };
    let Some(filename) = filename.to_str() else { return; };

    let Some(filename_components) = split_filepath_componentes(filename) else {
        return;
    };

    let l = filename_components.len();

    if let Ok(index) = filename_components[0].parse::<u32>()
        && index != header.index
    {
        // log::warn
    }

    if let Ok(energy) = parse_header_field(filename_components[l-6], "keV")
        && energy != header.energy
    {

    }

    if let Ok(temperature) = parse_header_field(filename_components[l-5], "K")
        && temperature != header.temperature
    {
        // log::warn
    }

    let sample = filename_components[1..l-7].join("_");

    if sample != header.sample {
        // log::warn
    }
}

fn import_spectrum_data(
    channel_data: &str
) -> Result<Spectrum, DATFormatError> {
    let channel_data = channel_data.split_whitespace();

    let mut spectrum: Vec<u32> = Vec::new();

    for x in channel_data {
        let x = x.parse::<u32>()?;
        spectrum.push(x);
    }

    Ok(Spectrum::from(spectrum))
}

fn import_palspectrum(filepath: &PathBuf, m: &mut Measurement) -> Result<PALSpectrum, ImportError> {
    let spectrum_file = File::open(filepath).map_err(
        |err| ImportError::FileSystemError {
            inner: err,
            path: filepath.into(),
        }
    )?;

    let mut reader = std::io::BufReader::new(spectrum_file);

    let mut header = String::new();
    reader.read_line(&mut header).map_err(
        |err| ImportError::FileSystemError {
            inner: err,
            path: filepath.into(),
        }
    )?;

    let mut channel_data = String::new();
    reader.read_to_string(&mut channel_data).map_err(
        |err| ImportError::FileSystemError {
            inner: err,
            path: filepath.into(),
        }
    )?;

    let header_info = get_header_info(&header, m)?;

    compare_filename_and_header(filepath, &header_info);

    let spectrum = import_spectrum_data(&channel_data)?;

    Ok(PALSpectrum::new(
        spectrum, "A".into(), (0., 1000. * header_info.bin_width)
    ))
}

fn import_params(filepath: &PathBuf, m: &mut Measurement) {
    let Ok(param_file) = File::open(filepath) else {
        // log::warn
        return;
    };

    let mut reader = std::io::BufReader::new(param_file);

    let mut content = String::new();
    let Ok(_) = reader.read_to_string(&mut content) else {
        return;
    };

    let lines = content.split("\n").map(|line| line.trim());

    for line in lines {
        if line.len() < 30 {
            continue;
        }

        let key = line[..29].trim();
        let value = line[29..].trim();

        m.metadata.insert(key.into(), value.into());
    }
}

fn import_ph_spectrum(filepath: &PathBuf) -> Option<DBSpectrum> {
    let Ok(param_file) = File::open(filepath) else {
        // log::warn
        return None;
    };

    let mut reader = std::io::BufReader::new(param_file);

    let mut spectrum_data = String::new();
    let Ok(_) = reader.read_to_string(&mut spectrum_data) else {
        return None;
    };

    let Ok(spectrum) = import_spectrum_data(&spectrum_data) else {
        return None;
    };

    Some(DBSpectrum::new(spectrum, "A".into(), (f64::NAN, f64::NAN), None))
}

pub fn import_meps_dat(filepath: &str) -> Result<Measurement, ImportError> {
    let datafiles = get_all_filenames(filepath)?;

    let dir = Path::new(filepath).parent().unwrap_or(Path::new(""));

    let mut m = Measurement::new();

    let p = import_palspectrum(&dir.join(datafiles.spectrum_file), &mut m)?;

    m.pals.insert("A".into(), p);

    import_params(&dir.join(datafiles.param_file), &mut m);

    if let Some(mut d1) = import_ph_spectrum(&dir.join(datafiles.start_phs_file)) {
        d1.detname = "Start".into();
        m.dbs.insert("Start".into(), d1);
    }

    if let Some(mut d2) = import_ph_spectrum(&dir.join(datafiles.stop_phs_file)) {
        d2.detname = "Stop".into();
        m.dbs.insert("Stop".into(), d2);
    }

    Ok(m)
}

File: kernels/Python/compositron/src/compositron/importers/meps_dat_importer.py
from dataclasses import dataclass
from pathlib import Path

import numpy as np

from compositron.dbs.dbspectrum import DBSpectrum

from ..pals import PALSpectrum
from ..core.measurement import Measurement
from ..core.utils import EnergyDetector, LinearCalibration, Spectrum, TimingDetectorPair, to_min_type_spectrum


@dataclass
class MePSFileBundle:
    spectrum_file: Path
    param_file: Path
    start_phs_file: Path
    stop_phs_file: Path


@dataclass
class HeaderInfo:
    index: int
    sample: str
    energy: float
    temperature: float
    bin_width: float


def split_filepath_components(filepath: Path) -> list[str]:
    return filepath.name.split("_")


def get_all_filenames(filepath: Path) -> MePSFileBundle:
    filename_components = split_filepath_components(filepath)
    dir = filepath.parent

    critical_idx = len(filename_components) - 3

    match filename_components[critical_idx]:
        case "PALS":
            begin = "_".join(filename_components[:critical_idx])
            end = "_".join(filename_components[critical_idx+1:])
            return MePSFileBundle(
                dir / "_".join(filename_components),
                dir / "_".join([begin, "Measurement_Parameter", end]),
                dir / "_".join([begin, "PHS_Ch0", end]),
                dir / "_".join([begin, "PHS_Ch1", end]),
            )
        case "Parameter":
            begin = "_".join(filename_components[:critical_idx-1])
            end = "_".join(filename_components[critical_idx+1:])
            return MePSFileBundle(
                dir / "_".join([begin, "PALS", end]),
                dir / "_".join(filename_components),
                dir / "_".join([begin, "PHS_Ch0", end]),
                dir / "_".join([begin, "PHS_Ch1", end]),
            )
        case "PALS":
            begin = "_".join(filename_components[:critical_idx-1])
            end = "_".join(filename_components[critical_idx+1:])
            return MePSFileBundle(
                dir / "_".join([begin, "PALS", end]),
                dir / "_".join([begin, "Measurement_Parameter", end]),
                dir / "_".join(filename_components),
                dir / "_".join([begin, "PHS_Ch1", end]),
            )
        case "PALS":
            begin = "_".join(filename_components[:critical_idx-1])
            end = "_".join(filename_components[critical_idx+1:])
            return MePSFileBundle(
                dir / "_".join([begin, "PALS", end]),
                dir / "_".join([begin, "Measurement_Parameter", end]),
                dir / "_".join([begin, "PHS_Ch0", end]),
                dir / "_".join(filename_components),
            )
        case _:
            raise ValueError("Invalid filepath")


def parse_header_field(field: str, expected_unit: str) -> float:
    l = len(field)
    u = len(expected_unit)

    value = float(field[:l-u])
    unit = field[l-u:]

    if unit == expected_unit:
        return value

    raise ValueError(
        f"Unexpected unit {unit} in spectrum file header "
        f"(expected {expected_unit})"
    )


def get_header_info(header: str, m: Measurement) -> HeaderInfo:
    header_fields = [field.strip() for field in header.split("_")]

    header_size = len(header_fields)

    if header_size < 6:
        raise ValueError("Spectrum header has unexpected format")

    index = int(header_fields[0])
    bin_width = parse_header_field(header_fields[header_size-1], "ns")
    temperature = parse_header_field(header_fields[header_size-3], "K")
    energy = parse_header_field(header_fields[header_size-4], "keV")
    sample = "_".join(header_fields[1:header_size-5])

    m.metadata["Sample"] = sample
    m.metadata["PositronImplantationEnergy"] = f"{energy}"
    m.metadata["SampleTemperature"] = f"{temperature}"

    return HeaderInfo(index, sample, energy, temperature, bin_width)


def compare_filename_and_header(filepath: Path, header: HeaderInfo) -> None:
    filename_components = split_filepath_components(filepath)

    l = len(filename_components)

    if int(filename_components[0]) != header.index:
        # TODO: warn
        pass

    if parse_header_field(filename_components[l-6], "keV") != header.energy:
        # TODO: warn
        pass

    if parse_header_field(filename_components[l-5], "K") != header.temperature:
        # TODO: warn
        pass

    if "_".join(filename_components[1:l-7]) != header.sample:
        # TODO: warn
        pass


def import_spectrum_data(spectrum_data: list[str]) -> Spectrum:
    return to_min_type_spectrum(
        np.array([int(x.strip()) for x in spectrum_data])
    )


def import_palspectrum(filepath: Path, m: Measurement) -> PALSpectrum:
    with open(filepath, "r") as file:
        lines = file.readlines()

    header = lines[0]
    spectrum_data = lines[1:]

    header_info = get_header_info(header, m)

    compare_filename_and_header(filepath, header_info)

    spectrum = import_spectrum_data(spectrum_data)

    detpair = TimingDetectorPair(
        name = "A",
        tcal = LinearCalibration(0, 1000 * header_info.bin_width),
        corrected_tcal = None,
        tres = None,
    )

    return PALSpectrum(spectrum, detpair)


def import_params(filepath: Path, m: Measurement):
    with open(filepath, "r") as file:
        lines = file.readlines()

    for line in lines:
        if len(line) < 30:
            continue

        key = line[:29].strip()
        value = line[29:].strip()

        m.metadata[key] = value


def import_ph_spectrum(filepath: Path) -> None | DBSpectrum:
    with open(filepath, "r") as file:
        spectrum_data = file.readlines()

    spectrum = import_spectrum_data(spectrum_data)

    detector = EnergyDetector(
        name = "",
        ecal = LinearCalibration(np.nan, np.nan),
        corrected_ecal = None,
        eres = None,
    )

    return DBSpectrum(spectrum, detector)


def import_meps_dat(filepath: str) -> Measurement:
    datafiles = get_all_filenames(Path(filepath))

    m = Measurement()

    p = import_palspectrum(datafiles.spectrum_file, m)

    m.pals["A"] = p

    import_params(datafiles.param_file, m)

    d1 = import_ph_spectrum(datafiles.start_phs_file)
    if d1 is not None:
        d1.detector.name = "Start"
        m.dbs["Start"] = d1

    d2 = import_ph_spectrum(datafiles.stop_phs_file)
    if d2 is not None:
        d2.detector.name = "Stop"
        m.dbs["Stop"] = d2

    return m
